util/cpuprofile/pprof_api.go

```diff
@@ -0,0 +1,204 @@
+	sec, err := strconv.ParseInt(r.FormValue("seconds"), 10, 64)
+	if sec <= 0 || err != nil {
+		sec = 30
+	}
+
+	if durationExceedsWriteTimeout(r, float64(sec)) {
+		serveError(w, http.StatusBadRequest, "profile duration exceeds server's WriteTimeout")
+		return
+	}
+
+	// Set Content Type assuming StartCPUProfile will work,
+	// because if it does it starts writing.
+	w.Header().Set("Content-Type", "application/octet-stream")
+	w.Header().Set("Content-Disposition", `attachment; filename="profile"`)
+
+	pc := NewPprofAPICollector()
+	err = pc.StartCPUProfile(w)
+	if err != nil {
+		serveError(w, http.StatusInternalServerError, "Could not enable CPU profiling: "+err.Error())
+		return
+	}
+	// TODO: fix me.
+	// This can be fixed by always starts a 1 second profiling one by one,
+	// but to aggregate (merge) multiple profiles into one according to the precision.
+	//  |<-- 1s -->|
+	// -|----------|----------|----------|----------|----------|-----------|-----> Background profile task timeline.
+	//                            |________________________________|
+	//       (start cpu profile)  v                                v (stop cpu profile)    // expected profile timeline
+	//                        |________________________________|                           // actual profile timeline
+	time.Sleep(time.Second * time.Duration(sec))
+	err = pc.StopCPUProfile()
+	if err != nil {
+		serveError(w, http.StatusInternalServerError, "Could not enable CPU profiling: "+err.Error())
+	}
+}
+
+const (
+	labelSQLDigest  = "sql_digest"
+	labelPlanDigest = "plan_digest"
+)
+
+// PprofAPICollector is a cpu profile consumer for Pprof API usage.
+type PprofAPICollector struct {
+	ctx       context.Context
+	cancel    context.CancelFunc
+	started   atomic.Bool
+	firstRead chan struct{}
+	wg        sync.WaitGroup
+
+	dataCh   ProfileConsumer
+	profiles []*ProfileData
+	writer   io.Writer
+}
+
+// NewPprofAPICollector returns a new NewPprofAPICollector.
+func NewPprofAPICollector() *PprofAPICollector {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &PprofAPICollector{
+		ctx:       ctx,
+		cancel:    cancel,
+		firstRead: make(chan struct{}),
+		dataCh:    make(ProfileConsumer, 1),
+	}
+}
+
+// StartCPUProfile is a substitute for the `pprof.StartCPUProfile` function.
+// You should use this function instead of `pprof.StartCPUProfile`.
+// Otherwise you may fail, or affect the TopSQL feature and pprof profile HTTP API .
+func (pc *PprofAPICollector) StartCPUProfile(w io.Writer) error {
+	if !pc.started.CAS(false, true) {
+		return errors.New("PprofAPICollector already started")
+	}
+	pc.writer = w
+	pc.wg.Add(1)
+	go goutil.WithRecovery(pc.readProfileData, nil)
+	return nil
+}
+
+// StopCPUProfile is a substitute for the `pprof.StopCPUProfile` function.
+func (pc *PprofAPICollector) StopCPUProfile() error {
+	if !pc.started.Load() {
+		return nil
+	}
+
+	// wait for reading least 1 profile data.
+	<-pc.firstRead
```

Comment:
Will deadlock happen?

Reply:
no, it should never happen.

Reply:
I do worry about this channel. For example, if there are problems for the global parallel collector, you may never receive any data? How about accepting the behavior that, if start and stop are called too close, there will be no data at all? Will it cause any problems?

Reply:
Ok, I alread add another timeout here.